crates/datasources/src/clickhouse/mod.rs

```diff
@@ -69,28 +75,54 @@ struct ClickhouseAccessState {
 
 impl ClickhouseAccessState {
     async fn connect(conn_str: &str) -> Result<Self> {
-        let pool = Pool::new(Options::new(Url::parse(conn_str)?).pool_min(1).pool_max(1));
+        let conn_str = Url::parse(conn_str)?;
+
+        // Verify that the conn_str has "clickhouse://" protocol.
+        if conn_str.scheme() != "clickhouse" {
+            return Err(ClickhouseError::String(format!(
+                "Expected url with scheme `clickhouse://`, got: `{}`",
+                conn_str.scheme()
+            )));
+        }
+
+        let addr = conn_str.clone(); // Let it take the address from original URL
+        let mut opts = Options::new(addr).pool_min(1).pool_max(1);
+
+        if let Some(mut path) = conn_str.path_segments() {
+            if let Some(database) = path.next() {
+                opts = opts.database(database);
+            }
+        }
+
+        let user = conn_str.username();
+        if !user.is_empty() {
+            opts = opts.username(user);
+        }
+
+        if let Some(password) = conn_str.password() {
+            opts = opts.password(password);
+        }
+
+        for (key, _val) in conn_str.query_pairs() {
+            if key.as_ref() == "s" || key.as_ref() == "secure" {
+                opts = opts.secure(true);
+            }
+        }
+
+        let pool = Pool::new(opts);
         let mut client = pool.get_handle().await?;
         client.ping().await?;
 
         Ok(ClickhouseAccessState { pool })
     }
 
-    async fn get_table_schema(&self, name: &str) -> Result<ArrowSchema> {
+    async fn get_table_schema(&self, table_ref: ClickhouseTableRef<'_>) -> Result<ArrowSchema> {
         let mut client = self.pool.get_handle().await?;
-        // TODO: Does clickhouse actually return blocks for empty data sets?
-        let mut blocks = client
-            .query(format!("SELECT * FROM {name} LIMIT 1"))
-            .stream_blocks();
 
-        let block = match blocks.next().await {
-            Some(block) => block?,
-            None => {
-                return Err(ClickhouseError::String(
-                    "unable to determine schema for table, no blocks returned".to_string(),
-                ))
-            }
-        };
+        let block = client
+            .query(format!("SELECT * FROM {table_ref} LIMIT 0"))
+            .fetch_all()
+            .await?;
```

Comment:
Just want to confirm, this is actually working with the `LIMIT 0` part? When I was trying to stream blocks, I wasn't getting anything. Is `fetch_all` behaving differently here?

Reply:
Yeah, this works. `fetch_all` does return a block, no matter how many rows.

Reply:
Nice, happy that this works.